File: src/server/controllers/browse.ts
import express from 'express';

import { getModelSkinOptions } from '@/lib/converter/character/utils';
import { getListFiles, registerListfileClearHook } from '@/lib/wow/listfile-cache';
import { isDirectListfileClient, tryInProcessListfileClient } from '@/lib/wow-data-client/direct-listfile';
import { FileEntry, wowDataClient } from '@/lib/wow-data-client/wow-data-client';
import {
  applyCascBuildCache, etagFromParts, matchNotModified, writeNotModified,
} from '@/server/utils/casc-cache';

let allFiles: FileEntry[] | null = null;
let modelFiles: FileEntry[] | null = null;
let textureFiles: FileEntry[] | null = null;

const m2WmoRegex = /\.(m2|wmo)$/i;
const badWmoRegex = /_([0-9]{3}|lod\d)\.wmo$/;
const textureRegex = /\.(blp|png|tga|dds)$/i;

export function ControllerBrowse(router: express.Router) {
  async function fetchAllFilesFromFullListfile() {
    allFiles = await getListFiles();
    allFiles = allFiles.filter((f) => !badWmoRegex.test(f.fileName));

    modelFiles = allFiles.filter((f) => m2WmoRegex.test(f.fileName))
      .sort((a, b) => a.fileName.localeCompare(b.fileName));
    console.log('Total M2/WMO files:', modelFiles.length);

    textureFiles = allFiles.filter((f) => textureRegex.test(f.fileName))
      .sort((a, b) => a.fileName.localeCompare(b.fileName));

    console.log('Total texture files:', textureFiles.length);
  }

  async function fetchBrowseIndex() {
    const start = Date.now();
    const inProcess = tryInProcessListfileClient();
    if (inProcess) {
      console.log('Building browse index from listfile...');
      const { models, textures } = await inProcess.collectBrowseFileIndex();
      modelFiles = models;
      textureFiles = textures;
      allFiles = null;
      console.log(`Collected ${models.length} models and ${textures.length} textures in ${((Date.now() - start) / 1000).toFixed(1)}s`);
      console.log('Total M2/WMO files:', modelFiles.length);
      console.log('Total texture files:', textureFiles.length);
      return;
    }
    if (isDirectListfileClient(wowDataClient)) {
      console.log('Building browse index from listfile...');
      try {
        const { models, textures } = await wowDataClient.collectBrowseFileIndex();
        modelFiles = models;
        textureFiles = textures;
        allFiles = null;
        console.log(`Collected ${models.length} models and ${textures.length} textures in ${((Date.now() - start) / 1000).toFixed(1)}s`);
        console.log('Total M2/WMO files:', modelFiles.length);
        console.log('Total texture files:', textureFiles.length);
        return;
      } catch (e) {
        console.warn('Direct browse index unavailable, falling back to full listfile copy:', (e as Error).message);
      }
    }
    console.log('Building browse index (full listfile copy)...');
    await fetchAllFilesFromFullListfile();
  }

  async function preloadBrowseIndex() {
    for (;;) {
      try {
        await wowDataClient.waitUntilReady();
        await fetchBrowseIndex();
        if ((modelFiles?.length ?? 0) > 0 || (textureFiles?.length ?? 0) > 0) return;
      } catch (e) {
        console.warn('browse index preload waiting for CASC:', (e as Error).message);
      }
      await new Promise<void>((resolve) => { setTimeout(resolve, 2000); });
    }
  }

  registerListfileClearHook(() => {
    allFiles = null;
    modelFiles = null;
    textureFiles = null;
    void preloadBrowseIndex();
  });

  void preloadBrowseIndex();

  // Search files in the WoW listfile; default to all m2 if no search
  router.get('/browse', async (req, res) => {
    try {
      if (!modelFiles && !textureFiles) {
        await fetchBrowseIndex();
      }
      const { q } = req.query as { q?: string };
      if (!q) {
        return res.status(400).json({ error: 'q is required' });
      }
      if (!['model', 'texture'].includes(q)) {
        return res.status(400).json({ error: 'q must be "model" or "texture"' });
      }

      let result: FileEntry[] = [];
      if (q === 'model') {
        result = modelFiles ?? [];
      } else if (q === 'texture') {
        result = textureFiles ?? [];
      }

      const buildKey = wowDataClient.cascInfo?.buildKey ?? '';
      const etag = etagFromParts('browse', buildKey, q, String(result.length));
      if (matchNotModified(req, etag)) {
        applyCascBuildCache(res, req, buildKey, etag);
        return writeNotModified(res, etag);
      }
      applyCascBuildCache(res, req, buildKey, etag);
      return res.json(result);
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });

  router.get('/browse/model-skins', async (req, res) => {
    try {
      const fileDataID = Number(req.query.fileDataID);
      if (!Number.isFinite(fileDataID) || fileDataID <= 0) {
        return res.status(400).json({ error: 'fileDataID is required' });
      }
      if (!modelFiles && !allFiles) {
        await fetchBrowseIndex();
      }
      const fileName = (modelFiles ?? allFiles ?? []).find((f) => f.fileDataID === fileDataID)?.fileName;
      if (!fileName) {
        return res.status(404).json({ error: 'Model not found' });
      }
      await wowDataClient.waitUntilReady();
      await wowDataClient.initModelCaches();
      const skins = await wowDataClient.getModelSkins(fileDataID);
      const options = await getModelSkinOptions(fileDataID, fileName, skins);
      return res.header('Cache-Control', 'no-store').json({ fileDataID, fileName, skins: options });
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });
}

File: golang/internal/server/api/browse.go
package api

import (
	"context"
	"log"
	"net/http"
	"regexp"
	"strconv"
	"sync"
	"time"

	"github.com/pqhuy98/wow-converter/internal/converter/character"
	"github.com/pqhuy98/wow-converter/internal/stringsort"
	"github.com/pqhuy98/wow-converter/internal/wow/casc"
	"github.com/pqhuy98/wow-converter/internal/wow/client"
)

var (
	m2WmoRegex   = regexp.MustCompile(`(?i)\.(m2|wmo)$`)
	badWmoRegex  = regexp.MustCompile(`(?i)_([0-9]{3}|lod\d)\.wmo$`)
	textureRegex = regexp.MustCompile(`(?i)\.(blp|png|tga|dds)$`)
)

type browseIndexes struct {
	mu           sync.Mutex
	modelFiles   []casc.ListfileEntry
	textureFiles []casc.ListfileEntry
	loading      bool
	loadErr      error
}

func (b *browseIndexes) load(ctx context.Context, c client.Client) error {
	b.mu.Lock()
	if b.modelFiles != nil {
		b.mu.Unlock()
		return nil
	}
	if b.loading {
		b.mu.Unlock()
		for {
			time.Sleep(50 * time.Millisecond)
			b.mu.Lock()
			if b.modelFiles != nil {
				b.mu.Unlock()
				return nil
			}
			if !b.loading {
				err := b.loadErr
				b.mu.Unlock()
				return err
			}
			b.mu.Unlock()
		}
	}
	b.loading = true
	b.mu.Unlock()

	err := b.build(ctx, c)

	b.mu.Lock()
	b.loading = false
	b.loadErr = err
	b.mu.Unlock()
	return err
}

func (b *browseIndexes) build(ctx context.Context, c client.Client) error {
	start := time.Now()
	var modelFiles, textureFiles []casc.ListfileEntry

	if direct, ok := c.(client.DirectListfileClient); ok {
		log.Printf("Building browse index from listfile...")
		var err error
		modelFiles, textureFiles, err = direct.CollectBrowseFileIndex(ctx)
		if err == nil {
			log.Printf("Collected %d models and %d textures in %.1fs", len(modelFiles), len(textureFiles), time.Since(start).Seconds())
			b.mu.Lock()
			b.modelFiles = modelFiles
			b.textureFiles = textureFiles
			b.mu.Unlock()
			log.Printf("Total M2/WMO files: %d", len(modelFiles))
			log.Printf("Total texture files: %d", len(textureFiles))
			return nil
		}
		log.Printf("Direct browse index unavailable (%v), falling back to full listfile copy", err)
	}

	if modelFiles == nil {
		log.Printf("Building browse index (full listfile copy)...")
		files, err := getListFiles(ctx, c)
		if err != nil {
			return err
		}
		log.Printf("Loaded %d listfile entries in %.1fs", len(files), time.Since(start).Seconds())

		filterStart := time.Now()
		filtered := make([]casc.ListfileEntry, 0, len(files))
		for _, f := range files {
			if badWmoRegex.MatchString(f.FileName) {
				continue
			}
			filtered = append(filtered, f)
		}

		modelFiles = make([]casc.ListfileEntry, 0, len(filtered)/4)
		textureFiles = make([]casc.ListfileEntry, 0, len(filtered)/2)
		for _, f := range filtered {
			if m2WmoRegex.MatchString(f.FileName) {
				modelFiles = append(modelFiles, f)
			} else if textureRegex.MatchString(f.FileName) {
				textureFiles = append(textureFiles, f)
			}
		}
		log.Printf("Filtered browse index in %.1fs", time.Since(filterStart).Seconds())

		sortStart := time.Now()
		sortBrowseEntries(modelFiles, textureFiles)
		log.Printf("Sorted browse index in %.1fs", time.Since(sortStart).Seconds())
	}

	b.mu.Lock()
	b.modelFiles = modelFiles
	b.textureFiles = textureFiles
	b.mu.Unlock()

	log.Printf("Total M2/WMO files: %d", len(modelFiles))
	log.Printf("Total texture files: %d", len(textureFiles))
	return nil
}

func (b *browseIndexes) snapshot(q string) []casc.ListfileEntry {
	b.mu.Lock()
	defer b.mu.Unlock()
	if q == "model" {
		return b.modelFiles
	}
	return b.textureFiles
}

func (b *browseIndexes) modelFileName(fileDataID int) string {
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, f := range b.modelFiles {
		if f.FileDataID == fileDataID {
			return f.FileName
		}
	}
	return ""
}

func sortBrowseEntries(modelFiles, textureFiles []casc.ListfileEntry) {
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		stringsort.SortBy(modelFiles, func(e casc.ListfileEntry) string { return e.FileName })
		wg.Done()
	}()
	go func() {
		stringsort.SortBy(textureFiles, func(e casc.ListfileEntry) string { return e.FileName })
		wg.Done()
	}()
	wg.Wait()
}

func waitForDataClient(ctx context.Context, c client.Client) {
	if err := c.WaitUntilReady(ctx); err != nil {
		log.Printf("browse preload: data client not ready: %v", err)
	}
}

func (b *browseIndexes) counts() (models, textures int) {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.modelFiles), len(b.textureFiles)
}

func registerBrowse(r Router, d *Deps) {
	indexes := &browseIndexes{}

	go func() {
		ctx := context.Background()
		waitForDataClient(ctx, d.Client)
		err := indexes.load(ctx, d.Client)
		if err != nil {
			log.Printf("browse index preload failed: %v", err)
			d.startup.browseFinished(false, 0, 0)
			return
		}
		models, textures := indexes.counts()
		d.startup.browseFinished(true, models, textures)
	}()

	r.Get("/browse", func(w http.ResponseWriter, req *http.Request) {
		if err := indexes.load(req.Context(), d.Client); err != nil {
			sendInternalError(w, err)
			return
		}
		q := req.URL.Query().Get("q")
		if q == "" {
			sendError(w, http.StatusBadRequest, "q is required")
			return
		}
		if q != "model" && q != "texture" {
			sendError(w, http.StatusBadRequest, `q must be "model" or "texture"`)
			return
		}
		w.Header().Set("Cache-Control", "public, max-age=60")
		sendJSON(w, http.StatusOK, indexes.snapshot(q))
	})

	r.Get("/browse/model-skins", func(w http.ResponseWriter, req *http.Request) {
		fileDataID, err := strconv.Atoi(req.URL.Query().Get("fileDataID"))
		if err != nil || fileDataID <= 0 {
			sendError(w, http.StatusBadRequest, "fileDataID is required")
			return
		}
		if err := indexes.load(req.Context(), d.Client); err != nil {
			sendInternalError(w, err)
			return
		}
		fileName := indexes.modelFileName(fileDataID)
		if fileName == "" {
			sendError(w, http.StatusNotFound, "Model not found")
			return
		}
		if err := d.Client.InitModelCaches(req.Context()); err != nil {
			sendInternalError(w, err)
			return
		}
		skins, err := d.Client.GetModelSkins(req.Context(), fileDataID)
		if err != nil {
			sendInternalError(w, err)
			return
		}
		options := character.GetModelSkinOptions(fileName, skins)
		w.Header().Set("Cache-Control", "no-store")
		sendJSON(w, http.StatusOK, map[string]any{
			"fileDataID": fileDataID,
			"fileName":   fileName,
			"skins":      options,
		})
	})
}
